packages/@react-stately/combobox/src/useComboBoxState.ts

```diff
@@ -79,27 +84,56 @@ export function useComboBoxState<T extends object>(props: ComboBoxStateProps<T>)
     items: props.items ?? props.defaultItems
   });
 
+  // This is only used if items are not controlled
+  let originalCollection = collection;
+
   let filteredCollection = useMemo(() => (
     // No default filter if items are controlled.
     props.items != null || !defaultFilter
       ? collection
       : filterCollection(collection, inputValue, defaultFilter)
   ), [collection, inputValue, defaultFilter, props.items]);
 
-  let triggerState = useMenuTriggerState(props);
-  let open = (focusStrategy?: FocusStrategy) => {
+  // Track what action is attempting to open the menu
+  let menuOpenTrigger = useRef('focus' as MenuTriggerAction);
+  let onOpenChange = (open: boolean) => {
+    if (props.onOpenChange) {
+      props.onOpenChange(open, open ? menuOpenTrigger.current : undefined);
+    }
+  };
+
+  let triggerState = useMenuTriggerState({...props, onOpenChange});
+  let open = (focusStrategy?: FocusStrategy, showAllItems?: boolean, trigger?: MenuTriggerAction) => {
     // Prevent open operations from triggering if there is nothing to display
-    if (allowsEmptyCollection || filteredCollection.size > 0) {
+    // Also prevent open operations from triggering if items are uncontrolled but defaultItems is empty, even if showAllItems is true.
+    // This is to prevent comboboxes with empty defaultItems from opening but allow controlled items comboboxes to open even if the inital list is empty (assumption is user will provide swap the empty list with a base list via onOpenChange returning `menuTrigger` manual)
+    if (allowsEmptyCollection || filteredCollection.size > 0 || (showAllItems && (originalCollection.size > 0 || props.items))) {
```

Comment:
Yep, but we still need this if statement to resolve to true so that `triggerState.open` gets called. Looking at it again, I can pull it out and make it  `...|| (showAllItems && originalCollection.size > 0) || props.items)` instead